rlkit/torch/sac/proto.py

```diff
@@ -16,26 +16,42 @@ def _product_of_gaussians(mus, sigmas_squared):
     sigmas_squared = torch.clamp(sigmas_squared, min=1e-7)
     sigma_squared = 1. / torch.sum(torch.reciprocal(sigmas_squared), dim=0)
     mu = sigma_squared * torch.sum(mus / sigmas_squared, dim=0)
-    return mu, torch.sqrt(sigma_squared)
+    return mu, sigma_squared
 
 
-def _mean_of_gaussians(mus, sigmas):
+def _mean_of_gaussians(mus, sigmas_squared):
     '''
     compute mu, sigma of mean of gaussians
     '''
     mu = torch.mean(mus, dim=0)
-    sigma = torch.sqrt(torch.mean(sigmas**2, dim=0))
-    return mu, sigma
+    sigma_squared = torch.mean(sigmas_squared, dim=0)
+    return mu, sigma_squared
+
+
+def _natural_to_canonical(n1, n2):
+    ''' convert from natural to canonical gaussian parameters '''
+    mu = -0.5 * n1 / n2
+    sigma_squared = -0.5 * 1 / n2
+    return mu, sigma_squared
+
+
+def _canonical_to_natural(mu, sigma_squared):
+    ''' convert from canonical to natural gaussian parameters '''
+    n1 = mu / sigma_squared
+    n2 = -0.5 * 1 / sigma_squared
+    return n1, n2
 
 
 class ProtoAgent(nn.Module):
 
     def __init__(self,
+                 env,
                  latent_dim,
                  nets,
                  **kwargs
     ):
         super().__init__()
+        self.env = env
```

Comment:
I don't feel too strongly, but I don't think the agent should have access to the env here. If all we're doing is calling sparsify reward, it's probably cleaner to just store it as an extra field in the replay buffer. It also just seems pretty hacky to me that the sparse reward is just a function of the original reward.

Reply:
I agree - we can do what I did when I was trying to make the MAESN envs work - stuff the sparse reward into `env_infos` (adding another field to the replay buffer is a pain).

Reply:
see `9aaf099`